core/server/worker/src/main/java/alluxio/worker/block/annotator/DefaultBlockIterator.java

```diff
@@ -0,0 +1,420 @@
+import alluxio.collections.Pair;
+import alluxio.worker.block.AbstractBlockStoreEventListener;
+import alluxio.worker.block.BlockMetadataManager;
+import alluxio.worker.block.BlockStoreEventListener;
+import alluxio.worker.block.BlockStoreLocation;
+import alluxio.worker.block.meta.StorageDir;
+import alluxio.worker.block.meta.StorageTier;
+
+import com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * The default {@link BlockIterator} implementation that integrates with the
+ * {@link BlockMetadataManager}.
+ */
+public class DefaultBlockIterator implements BlockIterator {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultBlockIterator.class);
+
+  /** Map of sorted block set collections per directory. */
+  private final HashMap<BlockStoreLocation, SortedBlockSet<BlockSortedField>> mPerDirOrderedSets;
+
+  /** Used to update total order for offline sorting schemes. */
+  private final ConcurrentHashSet<BlockStoreLocation> mUnorderedLocations;
+
+  /** Configured block annotator class. */
+  private final BlockAnnotator mBlockAnnotator;
+
+  /** Underlying meta manager. */
+  private final BlockMetadataManager mMetaManager;
+
+  /** Block store delegate listener. */
+  private BlockStoreEventListener mListener;
+
+  /**
+   * Creates default block iterator instance.
+   *
+   * @param metaManager meta manager
+   * @param blockAnnotator block annotator
+   */
+  public DefaultBlockIterator(BlockMetadataManager metaManager, BlockAnnotator blockAnnotator) {
+    mMetaManager = metaManager;
+    mBlockAnnotator = blockAnnotator;
+
+    mPerDirOrderedSets = new HashMap<>();
+    mUnorderedLocations = new ConcurrentHashSet<>();
+    mListener = new Listener();
+
+    initialize();
+  }
+
+  /**
+   * Initializes with the existing blocks.
+   */
+  private void initialize() {
+    // Initialize sets per location.
+    for (StorageTier tier : mMetaManager.getTiers()) {
+      for (StorageDir dir : tier.getStorageDirs()) {
+        mPerDirOrderedSets.put(dir.toBlockStoreLocation(), new SortedBlockSet());
+        mUnorderedLocations.add(dir.toBlockStoreLocation());
+      }
+    }
+
+    // Initialize with existing items.
+    for (StorageTier tier : mMetaManager.getTiers()) {
+      for (StorageDir dir : tier.getStorageDirs()) {
+        BlockStoreLocation dirLocation = dir.toBlockStoreLocation();
+        for (long blockId : dir.getBlockIds()) {
+          blockUpdated(blockId, dirLocation);
+        }
+      }
+    }
+  }
+
+  /**
+   * Called by block-store event callbacks to update a block.
+   */
+  private void blockUpdated(long blockId, BlockStoreLocation location) {
```

Comment:
For block updated and block removed, do we need to notify the block annotator?

Reply:
For updates, we need the annotator to give the block a new annotated value.
For removes, no need to invoke the annotator.